Incorrect result for query on date and timestamp predicate in partition table scenario

24nishant opened the pull request:
Problem:

ORCA generates incorrect results for a query on partition table with date and timestamp predicates.

Setup :

create tables
create table public.test
(datedday date)
WITH (
appendonly=false
)
PARTITION BY RANGE(datedday)
(
PARTITION pn_20221022 START ('2022-10-22'::date) END ('2022-10-23'::date),
PARTITION pn_20221023 START ('2022-10-23'::date) END ('2022-10-24'::date),
DEFAULT PARTITION pdefault
);

load data

insert into public.test(datedday)
select ('2022-10-22'::date)
union
select ('2022-10-23'::date);

query
select max(datedday) from public.test where datedday='2022-10-23' or datedday=('2022-10-23'::date -interval '1 day');
Sol:

ORCA retained the information of the functions used to access the predicate value and used it to access the other predicate also. For example, if the first predicate is of type 'date', then it used the same function to access the other predicate.
This caused the error, as in the query stated above, both predicate are of different type.
To solve the bug, the retention of old functions was removed.


**_NOTE: 
1. This PR addresses bug raised in https://github.com/greenplum-db/gpdb/issues/14303 

2. This PR is in continuation to old PR ref https://github.com/greenplum-db/gpdb/pull/14408
The old PR (14408) was closed due to rebasing error.  All the comments in the old PR are addressed here._**

24nishant commented:
To check the impact of updated code on dynamic partition elimination. I found that in case of dynamic partition elimination, the updated functions are never called.

I checked it in 2 ways,

By setting debug points : Found breakpoint never triggered
By running the same query on 6X_STABLE version also. The answer was same & correct for both 6X_STABLE and updated code for this PR.
I ran following queries to check the performance.

A.     SELECTION FROM TABLE (STATIC PARTITION ELIMINATION) – Access new fix code
 
1.     select count() from part2_date where  (part2_date.b=('2020-07-01'::date -interval '0 day') or part2_date.b='2020-08-01');
2.     select count() from part2_timestamp where  (part2_timestamp.b=('2020-07-01'::date -interval '0 day') or part2_timestamp.b='2020-08-01');
 
 
B.     JOIN WITH CONDITION (DYNAMIC PARTITION ELIMINATION) – Does not access new fix code
 
3.      select count() from bar,part2_date where (bar.b=part2_date.b) and ((part2_date.b=('2020-07-01'::date -interval '0 day') )or (part2_date.b='2020-08-01'));
4.     select count() from bar,part2_timestamp where (bar.b=part2_timestamp.b) and ((part2_timestamp.b=('2020-07-01'::date -interval '0 day') )or (part2_timestamp.b='2020-08-01'));
 
C.     JOIN WITHOUT CONDITION  – Does not access new fix code
 
5.     select count () from bar inner join part2_timestamp on (bar.b=part2_timestamp.b);   
6.     select count () from bar inner join part2_time on (bar.b=part2_time.b);   
 
 
D.     CROSS PRODUCT WITH CONDITION (STATIC PARTITION ELIMINATION) – Access new fix code
 
7.     explain select count() from bar,part2_date where (bar.b='2020-07-01' or bar.b='2020-08-01' ) and (part2_date.b=('2020-07-01'::date -interval '0 day') or part2_date.b='2020-08-01');   
8.     explain select count() from bar,part2_timestamp where (bar.b='2020-07-01' or bar.b='2020-08-01' ) and (part2_timestamp.b=('2020-07-01'::date -interval '0 day') or part2_timestamp.b='2020-08-01');   

Please check my findings in the attached document.
[Gpqp97.14Nov22.pdf](https://github.com/greenplum-db/gpdb/files/10089913/Gpqp97.14Nov22.pdf)

After this analysis, I believe that, the updated function is not called during dynamic partition elimination, thus the performance for dynamic partitioning should remain the same.

24nishant commented:
To check if the update for this PR (GPQP_97_6X_STABLE_FINAL) has any impact on the existing, static partition performance, i tried following 2 queries on a range partitioned table on 6X_STABLE and GPQP_97_6X_STABLE_FINAL(this update). The average execution and planning time for the 
queries was found similar for the both versions.

--------------------------------
SETUP
--------------------------------
create table foo_p(colm1 int,colm2 int) partition by range(colm1) (start(1) end(2000) every (2), default partition pdefault);
insert into foo_p select colm1, colm2 from generate_series(1,2000) colm1, generate_series(1,3) colm2;
Query 1: explain analyse select count(*) from foo_p where colm2=1;
Query 2: explain analyse select count(*) from foo_p where colm2=1 or colm2=2;

--------------------------------
RESULTS
--------------------------------
Please check the attached document. 
[Static Partition Results.pdf](https://github.com/greenplum-db/gpdb/files/10090463/Static.Partition.Results.pdf)

gpopt reviewed (commented):
Since the code change is not on Orca side, I would recommend not adding mdp tests. Just ICW test is ok.

24nishant commented:
> Since the code change is not on Orca side, I would recommend not adding mdp tests. Just ICW test is ok.

I have updated the code, and removed the mdp tests.

DevChattopadhyay commented on src/backend/cdb/cdbpartition.c:
```diff
@@ -4036,12 +4036,21 @@ selectListPartition(PartitionNode *partnode, Datum *values, bool *isnull,
 	PartitionListState *ls;
-
+	int	natts = partnode->part->parnatts;
 	if (accessMethods && accessMethods->amstate[partnode->part->parlevel])
```
Just for clarification can you please give an example to show in which case the if block will be executed and in which case the else block will be executed.

24nishant replied:
The "if-else" loop will be executed in a query where we have filter condition with different datatypes. Consider the following query (for query set up plz ref partition.sql) :

`select max(d) from test_listPartition where d='2022-10-23' or d=('2022-10-23'::date -interval '1 day');`

In this case, we have two datatypes in the filter condition, namely (1) Date, (2) Timestamp.
When the execution reaches, for the first time at the "if-else " loop, the else condition is executed first , the access methods are initialised. Based on this, the partition is selected later in the processing. 

Now when the execution return to this function to select partition in accordance with second filter condition, then because we saved the execution state in the last iteration, this time "IF" conditions executes.

DevChattopadhyay commented on src/backend/cdb/cdbpartition.c:
```diff
@@ -4036,12 +4036,21 @@ selectListPartition(PartitionNode *partnode, Datum *values, bool *isnull,
+		 */
+		for (int j=0;j< natts;j++)
+		{
+			ls->eqinit[j]=false;
```
What does the boolean value in eqinit actually signifies.

gpopt replied:
Please answer Dev's question.

24nishant replied:
To the best of my understanding "eqinit" is a flag, which tracks if we have selected the function for reading the filter values. Once the function is selected, "eqinit" are set to "true", so that when the execution reaches at this point again, we don't spend time in function selection. We simply check, if "eqint" is true, and proceed.

THANATOSLAVA commented:
Very interesting discovery. Thanks for the analysis.

My first impression was the update made `eqinit` redundant. Cause every time we set it `true` on line 4149 to indicate the function has been initialized, we reset it on line 4149 so the function gets initialized again.

Then I thought `eqinit` is probably redundant in `PartitionListState` to begin with. Without it, we should still be able to tell if the function has been initialized by checking if `fn_oid` equals `InvalidOid`.

So instead of setting/resetting `eqinit`, maybe we can do something like this, to avoid blind function reuse.
```c
Oid opfuncid = get_opfuncid_by_opname(opname, lhstypid, rhstypid);
if (ls->eqfuncs[i].fn_oid != opfuncid) {
    fmgr_info(opfuncid, &(ls->eqfuncs[i]));
}
``` 
Or, we can retain the previous `lhstypid` and `rhstypid`. If both are the same, we can reuse the function.

Similarly in `selectRangePartition`, batch reseting `fn_oid` might not be the best solution. Cause there's high chance that the previous function can be reused.

gpopt commented on src/backend/cdb/cdbpartition.c:
```diff
@@ -4036,12 +4036,21 @@ selectListPartition(PartitionNode *partnode, Datum *values, bool *isnull,
+		 *Setting the following condition, to ensure correct access function is selected
+		 *every time in a scenario with multiple predicates of different type in a partition table query.
+		 */
+		for (int j=0;j< natts;j++)
+		{
+			ls->eqinit[j]=false;
```
Code format. Also update the comment, add space after `*`, make sure the line width is shorter than 80 chars.
```c
		for (int j = 0; j < natts; j++)
		{
			ls->eqinit[j] = false;
```

24nishant replied:
I have updated in the observations in the latest commit.

gpopt commented on src/backend/cdb/cdbpartition.c:
```diff
@@ -4407,7 +4429,7 @@ selectRangePartition(PartitionNode *partnode, Datum *values, bool *isnull,
 		if (partnode->part->parlevel == 0)
 		{
-			int			i = 0;
+			int     i = 0;
```
Unneeded change

24nishant replied:
I have corrected in the latest commit.

gpopt commented on src/backend/cdb/cdbpartition.c:
```diff
@@ -4372,12 +4381,25 @@ selectRangePartition(PartitionNode *partnode, Datum *values, bool *isnull,
 	{
-		int			natts = partnode->part->parnatts;
+
 
```
Remove empty lines

24nishant replied:
Corrected in the latest commit.

24nishant commented:
> Very interesting discovery. Thanks for the analysis.
> 
> My first impression was the update made `eqinit` redundant. Cause every time we set it `true` on line 4149 to indicate the function has been initialized, we reset it on line 4049 so the function gets initialized again.
> 
> Then I thought `eqinit` is probably redundant in `PartitionListState` to begin with. Without it, we should still be able to tell if the function has been initialized by checking `fn_oid == InvalidOid`.
> 
> So instead of setting/resetting `eqinit`, maybe we can do something like this, to avoid blind function reuse.
> 
> ```c
> Oid opfuncid = get_opfuncid_by_opname(opname, lhstypid, rhstypid);
> if (ls->eqfuncs[i].fn_oid != opfuncid) {
>     fmgr_info(opfuncid, &(ls->eqfuncs[i]));
> }
> ```
> 
> Or, we can retain the previous `lhstypid` and `rhstypid`. If both are the same, we can reuse the function.
> 
> Similarly in `selectRangePartition`, batch reseting `fn_oid` might not be the best solution. Cause there's high chance that the previous function can be reused.

@ THANATOSLAVA, thank you for reviewing the PR so closely.

I worked on your inputs, and checked the implementation again. I would like to explain the implementation with an example, it should help us understand, why resetting of eqinit is required and retaining of variable is not required. 

Consider the following query, for list partition scenario:

**Setup:**

```
create table test_listPartition (i int, d date)
     partition by list(d)
 (partition p1 values('2022-10-22'), partition p2 values('2022-10-23'),
 default partition pdefault  );
```

```
insert into test_listPartition values(1,'2022-10-22');
insert into test_listPartition values(2,'2022-10-23');
insert into test_listPartition values(3,'2022-10-24');
```

Note: we have 3 partitions: P1, P2 and pdefault

**Query:**

`select max(d) from test_listPartition where d='2022-10-23' or d=('2022-10-23'::date -interval '1 day');`


**Analysis**

For static partition selection, the partition which satisfies the filter condidition is 
selected in function:selectListPartition(,,,,,,) . This function is called from partition_rules_for_equality_predicate(,,,), where inside a for loop every equality filter is selected and its value is send for partition selection. 

In our example query, first d=('2022-10-23'::date -interval '1 day') is processed,  this results in a value d='2022-10-22'::timestamp, so a timestamp value goes for partition selection. In the selectListPartition(), inside nested for loops, all the partitions are selected one by one, and value of 'd' is checked if it lies in the list or not. If value is found, a 'break' statement is executed & partition is selected.

To read this timestamp value, a function is also selected, only one, when the selectListPartition() is called
for a selected equality filter from partition_rules_for_equality_predicate().The 'eqinit' flags helps in restricting recomputation of this function. 

In the example query, partition P1 is selected, then the control goes back to partition_rules_for_equality_predicate(), 
from it selectListPartition() is again called. In this itereation it comes, with value  d='2022-10-23'::date. 
A new function is found to read this value, and partition P2 is seleted after processing. 

I have also analysed, the timing for a 1000 partition table and 4000 partition table, and found that the average planning
and execution time for the 6X_Stable and updated version are similar.

On the similar lines, the logic for range partition scenario is implemented.

THANATOSLAVA commented:
@24nishant Thanks for the detailed and structured explanation. I learned a lot. Essentially function initialization is only performed for the comparison of the predicate against the first partition rule. After that, `eqinit` staying `true` avoids function reinitialization for all the other partitions.

gpopt commented on src/backend/cdb/cdbpartition.c:
```diff
@@ -4372,12 +4382,27 @@ selectRangePartition(PartitionNode *partnode, Datum *values, bool *isnull,
+		 * correct access function is selected
+		 * every time in a scenario with
+		 * multiple predicates of different type
+		 * in a query.
```
Comment width is too short, keep it same width as L4386

24nishant replied:
Corrected in the code and uploaded in commit #e8cd382

gpopt commented on src/backend/cdb/cdbpartition.c:
```diff
@@ -4036,13 +4036,23 @@ selectListPartition(PartitionNode *partnode, Datum *values, bool *isnull,
+		 * correct access function is selected
+		 * every time in a scenario with multiple
+		 * predicates of different type.
+		 */
```
Comment width is too short, keep it same width as L4040

24nishant replied:
Corrected in the code and uploaded in commit #e8cd382

gpopt commented on src/backend/cdb/cdbpartition.c:
```diff
@@ -4036,13 +4036,23 @@ selectListPartition(PartitionNode *partnode, Datum *values, bool *isnull,
+		 * every time in a scenario with multiple
+		 * predicates of different type.
+		 */
+		for (int j=0; j < natts; j++)
```
`j=0` -> `j = 0`
Add space before and after `=`

24nishant replied:
Corrected in the code and uploaded in commit #e8cd382

gpopt commented on src/backend/cdb/cdbpartition.c:
```diff
@@ -4036,13 +4036,23 @@ selectListPartition(PartitionNode *partnode, Datum *values, bool *isnull,
+		 */
+		for (int j=0; j < natts; j++)
+		{
+			ls->eqinit[j]=false;
```
format issue:
space before and after `=`

24nishant replied:
Corrected in the code and uploaded in commit #e8cd382

gpopt commented on src/backend/cdb/cdbpartition.c:
```diff
@@ -4372,12 +4382,27 @@ selectRangePartition(PartitionNode *partnode, Datum *values, bool *isnull,
+			rs->ltfuncs_direct[keyno].fn_oid = InvalidOid;
+			rs->lefuncs_inverse[keyno].fn_oid = InvalidOid;
+			rs->ltfuncs_inverse[keyno].fn_oid = InvalidOid;
+		}
```
This is duplicate with L4421~4426.

Can we move them after L4445 so that the code is shared?

gpopt commented on src/backend/cdb/cdbpartition.c:
```diff
@@ -4443,6 +4415,17 @@ selectRangePartition(PartitionNode *partnode, Datum *values, bool *isnull,
 		else
 			rs->rules = NULL;
 	}
+	/*
```
Add 1 more empty line before this line.

gpopt reviewed (approved):
+1 after addressing new comments and rebase on latest 6X_STABLE.
